core/commands/add.go

```diff
@@ -43,14 +52,38 @@ remains to be implemented.
 	Options: []cmds.Option{
 		cmds.OptionRecursivePath, // a builtin option that allows recursive paths (-r, --recursive)
 		cmds.BoolOption("quiet", "q", "Write minimal output"),
+		cmds.BoolOption(progressOptionName, "p", "Stream progress data"),
+	},
+	PreRun: func(req cmds.Request) error {
+		req.SetOption(progressOptionName, true)
+
+		sizeFile, ok := req.Files().(files.SizeFile)
+		if !ok {
+			// we don't need to error, the progress bar just won't know how big the files are
+			return nil
+		}
+
+		size, err := sizeFile.Size()
+		if err != nil {
+			// see comment above
+			return nil
+		}
+		log.Debugf("Total size of file being added: %v\n", size)
+		req.Values()["size"] = size
+
+		return nil
 	},
-	Run: func(req cmds.Request) (interface{}, error) {
+	Run: func(req cmds.Request, res cmds.Response) {
 		n, err := req.Context().GetNode()
 		if err != nil {
-			return nil, err
+			res.SetError(err, cmds.ErrNormal)
+			return
 		}
 
+		progress, _, _ := req.Option(progressOptionName).Bool()
```

Comment:
this should also respect the `-q` flag. if `-q`, no progress bar. may be worth here just setting it:

```
if quiet {
  progress = 
```